### PR TITLE
Store the most recent changes time

### PR DESCRIPTION
**Pull request by jespino**
This fixes this issue: https://mattermost.atlassian.net/browse/MM-55561

### CODE REVIEW

File: server/handlers/handlers.go [hunk 1]
```diff
@@ -101,6 +102,36 @@ func (ah *ActivityHandler) Start() {
...
+	// doStart is the meat of the activity handler worker
+	doStartLastActivityAt := func() {
+		for {
+			timer := time.NewTimer(5 * time.Minute)
+			select {
+			case <-timer.C:
+				ah.lastUpdateAtMap.Range(func(subscriptionID, lastUpdateAt any) bool {
+					if lastUpdateAt.(time.Time).Add(5 * time.Minute).After(time.Now()) {
+						if err := ah.plugin.GetStore().UpdateSubscriptionLastActivityAt(subscriptionID.(string), lastUpdateAt.(time.Time)); err != nil {
+							ah.plugin.GetAPI().LogWarn("Error storing the subscription last activity at", "error", err)
```
Comment (lieut-data) [inline server/handlers/handlers.go:114]: Can we include the `subscription_id` and `last_update_at` in the log?

File: server/handlers/handlers.go [hunk 2]
```diff
@@ -101,6 +102,36 @@ func (ah *ActivityHandler) Start() {
...
 	}
 
+	// doStart is the meat of the activity handler worker
+	doStartLastActivityAt := func() {
+		for {
+			timer := time.NewTimer(5 * time.Minute)
+			select {
+			case <-timer.C:
+				ah.lastUpdateAtMap.Range(func(subscriptionID, lastUpdateAt any) bool {
+					if lastUpdateAt.(time.Time).Add(5 * time.Minute).After(time.Now()) {
```
Comment (lieut-data) [inline server/handlers/handlers.go:112]: nit: I had to sit down and think about what this meant. Maybe if we expressed it as the following instead?
```go
if time.Now().Sub(lastUpdateAt) <= 5 * time.Minute
```

File: server/handlers/handlers.go [hunk 3]
```diff
@@ -101,6 +102,36 @@ func (ah *ActivityHandler) Start() {
...
+		}
+	}
+
+	if lastActivityAtMap, err := ah.plugin.GetStore().GetSubscriptionsLastActivityAt(); err != nil {
+		ah.plugin.GetAPI().LogError("Unable to load last activity subscription", "error", err)
+	} else {
+		for subscriptionID, lastActivityAt := range lastActivityAtMap {
+			ah.lastUpdateAtMap.Store(subscriptionID, lastActivityAt)
+		}
+	}
```
Comment (lieut-data) [inline server/handlers/handlers.go:133]: Can we get away with not populating the map in the first place? I'm thinking about the error case here where something goes wrong and the map starts out empty. The `Range` above will just no-op until the first event, after which the subscription will get updated the very next time the job runs.

File: server/handlers/handlers.go [hunk 4]
```diff
@@ -101,6 +102,36 @@ func (ah *ActivityHandler) Start() {
...
+						if err := ah.plugin.GetStore().UpdateSubscriptionLastActivityAt(subscriptionID.(string), lastUpdateAt.(time.Time)); err != nil {
+							ah.plugin.GetAPI().LogWarn("Error storing the subscription last activity at", "error", err)
+						}
+					}
+					return true
+				})
+			case <-ah.quit:
+				timer.Stop()
+				// we have received a signal to stop
+				return
```
Comment (lieut-data) [inline server/handlers/handlers.go:122]: Should we flush our cache to the database when the worker exits?

File: server/store/sqlstore/store.go [hunk 5]
```diff
@@ -730,6 +734,35 @@ func (s *SQLStore) UpdateSubscriptionExpiresOn(subscriptionID string, expiresOn
 	return nil
 }
 
+func (s *SQLStore) UpdateSubscriptionLastActivityAt(subscriptionID string, lastActivityAt time.Time) error {
+	query := s.getQueryBuilder().Update(subscriptionsTableName).Set("lastActivityAt", lastActivityAt.UnixMicro()).Where(sq.Eq{"subscriptionID": subscriptionID}, sq.Lt{"lastActivityAt": lastActivityAt.Add(-10 * time.Second).UnixMicro()})
+	_, err := query.Exec()
+	if err != nil {
+		return err
```
Comment (lieut-data) [inline server/store/sqlstore/store.go:741]: For the global subscription, I expect this `UPDATE` to fail periodically due to concurrent writes. Propose we wrap this update in a `READ COMMITTED` transaction that will have Postgres seamlessly retry (re-evaluating the `WHERE` conditions). (Also, I'm separately proposing that we make this plugin Postgres only.)

File: server/store/sqlstore/store.go [hunk 6]
```diff
@@ -730,6 +734,35 @@ func (s *SQLStore) UpdateSubscriptionExpiresOn(subscriptionID string, expiresOn
...
+	}
+	return nil
+}
+
+func (s *SQLStore) GetSubscriptionsLastActivityAt() (map[string]time.Time, error) {
+	query := s.getQueryBuilder().Select("subscriptionID, lastActivityAt").From(subscriptionsTableName).Where(sq.NotEq{"lastActivityAt": nil}, sq.NotEq{"lastActivityAt": 0})
```
Comment (lieut-data) [inline server/store/sqlstore/store.go:747]: nit: I find it easier to parse and later extend these when they're written as:
```go
	query := s.getQueryBuilder().
		Select("subscriptionID, lastActivityAt").
		From(subscriptionsTableName).
		Where(sq.NotEq{
			"lastActivityAt": nil, 
			"lastActivityAt": 0,
		})
```

**Review by lieut-data — CHANGES_REQUESTED** [review]
Nice! Laying the foundation for the next steps of using this data. A few comments below.

File: server/store/sqlstore/store.go [hunk 7]
```diff
@@ -730,6 +734,35 @@ func (s *SQLStore) UpdateSubscriptionExpiresOn(subscriptionID string, expiresOn
 	return nil
 }
 
+func (s *SQLStore) UpdateSubscriptionLastActivityAt(subscriptionID string, lastActivityAt time.Time) error {
```
[the remark this answers is not in the document]
Comment (jespino) [reply server/store/sqlstore/store.go:737]: Extend is, in my opinion a bit more confusing. We are not "extending" it, maybe "AdvanceSubscriptionLastActivityAt"? But in that case I would pass a "duration", and we don't pass a duration. I think the `Update` one ist he one that defines better the behavior in terms of parameters and internal behavior. I could consider acceptable "AdvanceSubscriptionLastActivityAtTo", and pass the date that we want to put there.

File: server/handlers/handlers.go (hunk 1 above)
Comment (jespino) [reply server/handlers/handlers.go:114]: sure

File: server/handlers/handlers.go (hunk 3 above)
Comment (jespino) [reply server/handlers/handlers.go:133]: Something to think about, I think the problem of not having that initialized is about not syncing properly the messages later.

File: server/handlers/handlers.go (hunk 4 above)
Comment (jespino) [reply server/handlers/handlers.go:122]: We could.

File: server/store/sqlstore/store.go (hunk 5 above)
Comment (jespino) [reply server/store/sqlstore/store.go:741]: This shouldn't happen concurrently for the same subscription.

File: server/handlers/handlers.go (hunk 3 above)
Comment (lieut-data) [reply server/handlers/handlers.go:133]: I'm assuming when we go to synchronize, we'll want to get the latest values from the database and not rely on the map, as we won't necessarily "see" changes that only get to other plugin instances.

File: server/store/sqlstore/store.go [hunk 8]
```diff
@@ -730,6 +734,35 @@ func (s *SQLStore) UpdateSubscriptionExpiresOn(subscriptionID string, expiresOn
...
+	query := s.getQueryBuilder().Update(subscriptionsTableName).Set("lastActivityAt", lastActivityAt.UnixMicro()).Where(sq.Eq{"subscriptionID": subscriptionID}, sq.Lt{"lastActivityAt": lastActivityAt.Add(-10 * time.Second).UnixMicro()})
```
[the remark this answers is not in the document]
Comment (jespino) [reply server/store/sqlstore/store.go:738]: I think is not even needed, I'm removing it. That comes from an intermediate state of the code where I followed a different approach to store the the last activity at.

File: server/handlers/handlers.go (hunk 3 above)
Comment (jespino) [reply server/handlers/handlers.go:133]: Yes, that's true, we can relay on database content, and that should be enough.

File: server/store/sqlstore/store.go (hunk 5 above)
Comment (jespino) [reply server/store/sqlstore/store.go:741]: Still, they are idempotent and 1 transaction per query, so I don't see any problem here. We can talk it later today.

Comment (lieut-data) [reply server/store/sqlstore/store.go:744]: As discussed offline, the default transaction isolation level is already `READ COMMITTED`, so we're all good 🎉

File: server/handlers/handlers.go (hunk 3 above)
Comment (lieut-data) [reply server/handlers/handlers.go:135]: Discussed offline, but agreed we can just remove this and keep it simpler 🎉

**Review by wiggin77 — APPROVED** [review]
LGTM :+1:

**Comment by jespino** [comment]
@lindalumitchell There is nothing user facing to test here. The idea is that any activity coming from MSTeams side should be generating updates on the database (storing the last activity date). For testing this I would suggest some smoke testing around creation channel subscriptions and verifying that messages flows correctly in both directions. Feel free to defer the testing after the merge, to do it as part of the regular testing. I can give it a quick round myself after or before the merge if you want.

**Review by lindalumitchell — APPROVED** [review]
Sounds like a good plan to me, thanks @jespino!

**Pull request merged**